### PR TITLE
Tagging specs

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

**Comment by jimrollenhagen** [comment]
So I know specs are supposed to be about business rules and all that... but this feature will need some relatively major architectural considerations... where do we want to talk about that? :)

**Comment by jimrollenhagen** [comment]
(leaving some notes in #150...)

**Comment by ghost** [comment]
@jimrollenhagen Let's talk about it in the issue. I agree on the architectural considerations, in fact I was expected this spec to be less use case design and more architecture...

File: doc/specs/tagging.md [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+
+Tagging allows items on the site to be associated with a topic, either broad or granular. Having 
+such a system in place allows users to find exactly what they're looking for with much greater
+ease.
+
+## Business rules
+
+* Tags must come in two options:
+  * Simple tagging, which allows for an unlimited number of tags.
+  * Hierarchal tagging, which allows for sub-tags.
```
Comment (ajax-wcd) [inline doc/specs/tagging.md]: How deep can they go?

File: doc/specs/tagging.md [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+* The number of tags must have a limit that is configurable on a case-by-case basis. (i.e. a limit
+  of 5 tags for a wiki article, or 2 tags for collages)
+  * In the case of hierarchal tagging, the following options must be configurable:
+    * Required or optional parent tags, minimum number of parent tags, or maximum number of parent tags.
+    * Required or optional child tags, minimum number of child tags, or maximum number of child tags.
+    * Selection of parent tags before child tags required or optional.
+* Each system that uses the tagging packge must have the option to force tags to be chosen from
+  a pre-defined set. (i.e. a forum limiting users to tags relevant to the forum)
+* Each system that uses the tagging package must have tag editing permissions specific to the system.
```
Comment (ghost) [inline doc/specs/tagging.md:28]: Permissions can be on a tag basis too, in other words a user can have access to only a subset of tags.

File: doc/specs/tagging.md [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+This applies if there is no pre-defined tag set.
+
+1. The user opens the tag manager.
+2. The user enters tags up to the set limit, or as many as the user would like if no limit
+   is set.
+3. The user closes the tag manager, and the tags are added to the item.
+
+**Non-Functional Requirements:**
+
+* Tags can be edited at any point during the creation of the item.
```
Comment (ghost) [inline doc/specs/tagging.md:81]: Implied.

File: doc/specs/tagging_technical.md [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+  * Get the taglist for an item
+  * Data: None
+* /%item%/:id/tags/:tag_id/upvote [PUT]
+  * Upvote a tag
+  * Data: None
+* /%item%/:id/tags/:tag_id/downvote [PUT]
+  * Downvote a tag
+  * Data: None
+
+### Global Tags
```
Comment (ghost) [inline doc/specs/tagging_technical.md:101]: uhhh, we don't have http requests in meteor. Applies thoughout spec

File: doc/specs/tagging_technical.md [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+## Preface
+
+The tagging system is part of the core of meteor-gazelle. Tagging allows users to add labels
+to their items, simplifying the way different sections of the site are searched. It allows
+for more fine-graned queries, limiting results to the desired item set.
+
+## Class Implementation
+
+* Class: TagManager - Attached to individual items (threads, wiki articles, etc.)
```
Comment (jimrollenhagen) [inline doc/specs/tagging_technical.md:11]: Can we just like, write the method signatures and docstrings or whatever JS calls them, in code blocks? This is pretty impossible to read, doesn't give argument names, etc

File: doc/specs/tagging_technical.md [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    * Add a tag to the global blacklist
+  * Function: removeTagFromBlacklist(tag_id)
+    * Remove a tag from the global blacklist
+  * Function: renameTag(tag_id, new_name, type)
+    * Rename a tag
+    * Type is a string (whitelist or blacklist), indicating if the tag is on the whitelist or blacklist
+
+## Data Model
+
+* %item%_tags:
```
Comment (jimrollenhagen) [inline doc/specs/tagging_technical.md:65]: So a collection per item? This could get weird in code.

File: doc/specs/tagging_technical.md [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+    * tag_id: string, the tag ID
+    * parent_tag: boolean, only created if heirarchal tagging is chosen. true if parent tag,
+      false if child tag
+    * votes: int, only created if voting is enabled. Contains the net number of votes
+    * added_by: string, the ID of the user that added the tag
+    * added_at: string, the timestamp of when the tag was added
+* global_tags:
+  * tag: string, the tag name
+* tag_blacklist:
+  * tag: string, the tag name
```
Comment (jimrollenhagen) [inline doc/specs/tagging_technical.md:77]: Assuming a collection per item, global blacklist seems odd.

File: doc/specs/tagging_technical.md (hunk 4 above)
Comment (jimrollenhagen) [inline doc/specs/tagging_technical.md:101]: +1, maybe call out the Meteor.methods the client sees?

File: doc/specs/tagging_technical.md [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+
+* %item%_tags:
+    * %item%_id: int, the item ID (thread ID, wiki article ID, etc.)
+    * tag: string, the tag name
+    * tag_id: string, the tag ID
+    * parent_tag: boolean, only created if heirarchal tagging is chosen. true if parent tag,
+      false if child tag
+    * votes: int, only created if voting is enabled. Contains the net number of votes
+    * added_by: string, the ID of the user that added the tag
+    * added_at: string, the timestamp of when the tag was added
```
Comment (jimrollenhagen) [inline doc/specs/tagging_technical.md:73]: convention is created_by, created_at

File: doc/specs/tagging_technical.md [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+    * Type is a string (whitelist or blacklist), indicating if the tag is on the whitelist or blacklist
+
+## Data Model
+
+* %item%_tags:
+    * %item%_id: int, the item ID (thread ID, wiki article ID, etc.)
+    * tag: string, the tag name
+    * tag_id: string, the tag ID
+    * parent_tag: boolean, only created if heirarchal tagging is chosen. true if parent tag,
+      false if child tag
```
Comment (jimrollenhagen) [inline doc/specs/tagging_technical.md:70]: Seems like this should be an ID pointing at another tag... if we do this heirarchy thing it should be arbitrary tree structures IMO

File: doc/specs/tagging_technical.md (hunk 8 above)
Comment (rhomes-wcd) [inline doc/specs/tagging_technical.md:73]: Following the meteor style guide and the mongo guidelines that will be merged, we want collections, fields, and variable identifiers to be camelCased.

**Pull request closed without merging**